Use url.resolve to support relative URLs with path.

boushley opened the pull request:
I had a bug where this wouldn't download playlists that use relative paths and were in a directory instead of the root of the host. This uses the built in [`url.resolve`](https://nodejs.org/api/url.html#url_url_resolve_from_to) to simplify URL parsing.

nmrony commented:
@boushley Many many thanks for sending a pull request. I will review this. Can you please check why the test cases are failing? I really appreciate your. I am in office now. I will jump into it as soon I reach home and manage some time for it

boushley commented:
Good catch. I didn't look at the tests. I removed a fair number of them since they are no longer applicable. I added some new tests to ensure we continue to handle the many different types of URLs.

nmrony commented:
@boushley Is there any problem in `buildURL` function to support relative paths? `buildURL` function was written to add relative relative path support. Please consider an use case, suppose if  master playlist file points its variants to different domain then I think case will fail. Please do not remove any existing test case please comment it out with some words if you think they are not appropriate. If relative path is not parsing well then I think we need to look into `buildURL` function. I am busy with family for [Eid Festival](https://en.wikipedia.org/wiki/Eid_al-Fitr). If you are not that busy can you please look into it? I will be grateful to you. I will dive into it after Eid.  I am feeling honored having you beside me :+1:

boushley commented:
`buildURL` failed to handle URLs where there were path elements besides the manifest. So it handled urls like `http://www.example.com/foo.m3u8` but did not handle `http://www.example.com/foo/bar/baz.m3u8`. The stream manifests or transport stream URLs would end up on the root instead of in the `foo/bar` path.

What's more the `buildURL` function is unnecessary since the built in `url` module has the [`url.resolve`](https://nodejs.org/api/url.html#url_url_resolve_from_to) method that performs this functionality.

My proposal is to remove the `buildURL` method entirely and use the more robust build in `url.resolve`. As such the test cases around `buildURL` are invalid, and I added better tests around manifest parsing that test urls being resolved correctly.

Enjoy the time with family.

nmrony commented:
Cool!!! @boushley You are really awesome. though I am in vacation, I do not want you wait to test and merge this. Great job buddy!!! Have no words to express my gratitude. We need your help to make this more robust. :tada: :tada: :sparkles:. I will be more than happy adding you as a collaborator of this project. What do you say? :stuck_out_tongue_winking_eye:

boushley commented:
Thanks for the offer @nmrony 

I'm not sure I can contribute any sort of regular time to this, but if you make me a contributor it would help with any issues I might have later.

nmrony commented:
@boushley You can consider it as a fun side project. Contribute whenever you manage time. Welcome to  the party  :tada: :tada: :balloon:
